### DISCUSSION TITLE
/login endpoint is throwing 403 and frontend only the login interface requests from the frontend to the backend?

### DISCUSSION THREAD

**Discussion by gweesin** [Project-Books/book-project#959] [Q&A]
I run the frontend and backend program alone(main branch). But the backend postman cannot request successfully, and the frontend cannot display the page correctly.

## 1. about frontend

this is my frontend package.json(I just delete the proxy field):

```
{
  "name": "book-project",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@apollo/client": "^3.4.16",
    "@material-ui/core": "^4.11.2",
    "@material-ui/icons": "^4.11.2",
    "@material-ui/lab": "^4.0.0-alpha.60",
    "@material-ui/styles": "^4.11.2",
    "@testing-library/jest-dom": "^5.11.4",
    "@testing-library/react": "^11.1.0",
    "@testing-library/user-event": "^12.1.10",
    "@types/jest": "^26.0.15",
    "@types/node": "^12.0.0",
    "@types/react": "^17.0.0",
    "@types/react-dom": "^17.0.0",
    "bootstrap": "^5.1.1",
    "graphql": "^15.6.1",
    "react": "^17.0.1",
    "react-bootstrap": "^2.0.0-rc.0",
    "react-dom": "^17.0.1",
    "react-router": "^5.2.0",
    "react-router-dom": "^5.2.0",
    "react-scripts": "^5.0.0",
    "typescript": "^4.0.3",
    "web-vitals": "^0.2.4",
    "zxcvbn": "^4.4.2"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "lint": "eslint . --ext=.ts,.tsx",
    "tsc": "tsc"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@types/react-router-dom": "^5.1.7",
    "@types/zxcvbn": "^4.4.0",
    "@typescript-eslint/eslint-plugin": "^4.14.1",
    "@typescript-eslint/parser": "^4.14.1",
    "eslint": "^7.18.0",
    "eslint-plugin-react": "^7.22.0",
    "eslint-plugin-react-hooks": "^4.2.1-rc.0-next-419ccc2b1-20220217"
  }
}
```

However, the frontend only the login interface requests from the frontend to the backend. other port of requests is 3000(the frontend port):

![image](https://user-images.githubusercontent.com/42909374/158059766-a66a7587-b652-47a4-9bc5-c922d11d4864.png)

Very confused that in addition to the `/login` interface, all other requests are for port 3000.😭 like this: 

![image](https://user-images.githubusercontent.com/42909374/158059793-23f7f224-c436-4f90-9ffb-87d80eb3e2c2.png)

## 2. about backend

I try the [discussion843](https://github.com/Project-Books/book-project/discussions/843) but also many problems.

```
// in SecurityConfiguration.java
@Override
protected void configure(HttpSecurity http) throws Exception {
    http.cors()
        .and()
        .csrf().disable()
        .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS)
        .and()
        .addFilter(new JwtUsernamePasswordAuthFilter(authenticationManager(), jwtConfig,
                secretKey))
        .addFilterAfter(new JwtTokenVerifier(secretKey, jwtConfig), JwtUsernamePasswordAuthFilter.class)
        .authorizeRequests()
        .antMatchers(HttpMethod.POST, Mappings.USER).permitAll()
        .anyRequest()
        .authenticated();
}
```

I just modify the config file(application.properties) like this:

```
server.port=8082

spring.jpa.hibernate.ddl-auto=validate
spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.PostgreSQLDialect
spring.datasource.url = jdbc:postgresql://localhost:5432/book_project_db
spring.datasource.username = postgres
spring.datasource.password = 123456

spring.data.rest.base-path=/api

application.jwt.secretKey=4&lTw7H@VWQtYC$3gh9DD9cReS7ehZ&tL9^wq@ODH@#8TCIZuRkcYE2%Qg$3X$!2
application.jwt.tokenPrefix=Bearer 
application.jwt.tokenExpirationAfterDays=7
```

but see the postman:

![image](https://user-images.githubusercontent.com/42909374/158059724-6b0b1053-e969-4050-99fb-a31fe9f951c4.png)

Thank you for your willingness to help me!!! 😃

**Comment by knjk04** [comment]
Hi @c-g-x, you shouldn't need to change anything to get it working. I noticed you changed the database settings. Does it work when you also run the database within a Docker container (and revert back to the database settings we have in the `main` branch)?

If you don't want to run the database in a Docker container, you'll need to manually create your own Postgres database (which we don't recommend because it's easier to help people out if they're using our setup).

The login endpoint requires the `POST` verb. See this guide for more information: https://project-books.github.io/development/how-to/backend-postman/

**Comment by gweesin** [comment]
well, @knjk04, the login request is `post`. This is my negligence... 
The reason why I didn't use docker and set up my own database is that I found that my docker reported an error when it was running, and I didn't know anything about docker before.

So, the postman send request successfully!

![image](https://user-images.githubusercontent.com/42909374/158062458-031d413c-0574-4648-a6f7-43198182114a.png)

But the frontend also can not request successfully(expect /login).

**Reply by knjk04** [reply]
What's the error on the frontend? 

Also, what was the error you were getting with Docker?

**Reply by gweesin** [reply]
## the frontend condition

the login request is normal:

![image](https://user-images.githubusercontent.com/42909374/158062896-811144e3-a7d6-479f-b42b-d1254b6fc350.png)

but other requests of port is 3000(the port of frontend), no 8080.

![image](https://user-images.githubusercontent.com/42909374/158062921-4f274a3c-89dd-4a61-b372-721d0b39cb75.png)

## the Docker condition

I'm reinstalling now

**Reply by gweesin** [reply]
if i reset the branch to the head of `main`, `yarn install` and `yarn start`, the console we can see:

![image](https://user-images.githubusercontent.com/42909374/158064356-3e47bd01-2195-4da1-81fc-cff38fce9a4e.png)

I think the reason is the proxy field in `package.json` is not exist.

and the page we can see:

![image](https://user-images.githubusercontent.com/42909374/158064423-a410d0d2-3f74-484a-8995-20760c3b0bc3.png)

maybe the problem of eslint-plugin-react-hooks version???

**Reply by knjk04** [reply]
Just to clarify, is this when the branch is exactly the same as `main`? I know you said before that you deleted the proxy line in `package.json`, so I just want to make sure

**Reply by gweesin** [reply]
right, the branch is exactly the same as `main`.
I install the docker again, now the start up by docker command has the same problem:
![image](https://user-images.githubusercontent.com/42909374/158383074-c7e581ef-e4e5-407b-a07d-89a24100e19d.png)
Nevertheless, I can still run the web page:
![image](https://user-images.githubusercontent.com/42909374/158383305-62e4d7d2-21ca-4d31-a893-02a89fe9d778.png)
I didn't respond to those parts in the red box. Why? I'm dizzy now😵

**Reply by knjk04** [reply]
@c-g-x Thanks for bringing up this error, I was able to replicate it locally. 

The error is here: https://github.com/Project-Books/book-project/blob/main/frontend/src/shared/http/HttpClient.ts#L40. Even when we correctly indent the ignore comment on the line above (line 39), it still doesn't ignore the use of `any`. I'm not sure why that is just yet, but we're looking into this!

**Reply by igiannopi** [reply]
Hey, I would like to start working on this project. However, I am wondering, reading the above comments, would be possible for me to install and run the project?
